src/main/scala/io/qbeast/spark/index/OTreeDataAnalyzer.scala

```diff
@@ -115,18 +122,21 @@ object DoublePassOTreeDataAnalyzer extends OTreeDataAnalyzer with Serializable {
         }
     }
 
-  private[index] def estimatePartitionCubeSize(
+  private[index] def estimateGroupCubeSize(
       desiredCubeSize: Int,
-      numPartitions: Int): Double = {
-    if (numPartitions > 0) {
-      val desiredPartitionCubeSize =
-        Math.ceil(desiredCubeSize.toDouble / numPartitions + 1) // Round to the next value
-      if (desiredPartitionCubeSize < minPartitionCubeSize) {
+      numPartitions: Int,
+      numElements: Long,
+      maxWorkingSetSize: Long): Double = {
+    val numGroups = Math.max(numPartitions, numElements / maxWorkingSetSize)
+    if (numGroups > 0) {
+      val groupCubeSize =
+        Math.ceil(desiredCubeSize.toDouble / numGroups + 1)
```

Comment:
Here the `groupCubeSize` is set as `desiredCubeSize / numGroups + 1`

In case we did not round this number, the next situation happens:

1. We index a dataframe of `10000` elements and we set the cubeSize to `10000`.
2. The groupCubeSize is `10000`
3. We end up having one cube with `9998` elements and another one with `2`. Because the CubeWeights only outputs a single cube with max weight 0.999.. and when storing the elements, the ones in the range [0.999, 1.0] end up in the children.

My question is: What is the behavior expected? To have 2 cubes or to store all the data in a single one if the cube size matches the size of the dataset?

@cugni @alexeiakimov any comment on this would be helpful!

Reply:
It is a target cube size, it is a "soft limit", or better a recommendation, we do not have to comply strictly with it.